ops/ccs-ops-misc/synthetic-data/scripts/generate-characteristics-file.py

```diff
@@ -0,0 +1,166 @@
+        hospice_data = get_table_count("hospice_claims", bene_id_start, bene_id_end, db_string)
+        inpatient_data = get_table_count("inpatient_claims", bene_id_start, bene_id_end, db_string)
+        outpatient_data = get_table_count("outpatient_claims", bene_id_start, bene_id_end, db_string)
+        snf_data = get_table_count("snf_claims", bene_id_start, bene_id_end, db_string)
+        pde_data = get_table_count("partd_events", bene_id_start, bene_id_end, db_string)
+    except BaseException as err:
+        print(f"Unexpected error while running queries: {err}")
+        print("Returning with exit code 1")
+        sys.exit(1)
+    
+    ## synthesize data into final rows
+    final_data_rows = put_data_into_final_rows(bene_data, carrier_data, dme_data, hha_data, hospice_data, inpatient_data, outpatient_data, snf_data, pde_data)
+    
+    ## Write csv to filesystem + header
+    filePath = output_path + 'characteristics.csv'
+    print("Writing final csv...")
+    try:
+        with open(filePath, 'w') as f:
+            writer = csv.writer(f)
+            writer.writerow(header)
+            writer.writerows(final_data_rows)
+            num_rows = len(final_data_rows)
+            print(f"Wrote out {num_rows} to {filePath}")
+    except IOError as err:
+        print(f"IOError while opening/writing csv: {err}")
+        print("Returning with exit code 1")
+        sys.exit(1)
+    except BaseException as err:
+        print(f"Unexpected error while opening/writing csv: {err}")
+        print("Returning with exit code 1")
+        sys.exit(1)
+    
+    print("Returning with exit code 0 (No errors)")
+    sys.exit(0)
+    
+
+def get_bene_data(bene_id_start, bene_id_end, db_string):
+    """
+    Gets the initial data from the beneficiary table including the beneficiary id, 
+    mbi, and a concatenated list of contract numbers.
+    """
+    
+    query = f"SELECT bene_id, mbi_num, concat_ws(',', ptd_cntrct_jan_id, ptd_cntrct_feb_id,ptd_cntrct_mar_id,ptd_cntrct_apr_id,ptd_cntrct_may_id,ptd_cntrct_jun_id,"\
+        f" ptd_cntrct_jul_id, ptd_cntrct_aug_id, ptd_cntrct_sept_id, ptd_cntrct_oct_id, ptd_cntrct_nov_id, ptd_cntrct_dec_id) as \"Part D Contract Number\""\
+        f" FROM public.beneficiaries WHERE bene_id <= {bene_id_start} and bene_id > {bene_id_end} order by bene_id desc"
+        
+    print(f"Starting query for bene data...");
+    raw_query_response = _execute_query(db_string, query)
+    rows = len(raw_query_response)
+    print(f"Got {rows} results from bene data query.");
+    return raw_query_response
+    
+def get_table_count(table_name, bene_id_start, bene_id_end, db_string):
+    """
+    Gets the table count for each bene in the specified range for the specified
+    database, and returns a dictionary with the bene id as the key and the
+    table count as the value.
+    """
+    
+    query = "SELECT t1.bene_id, count(t2.*)"\
```

Comment:
Curious on this logic; do we even need to query the beneficiaries table here. All the parent claims tables will a) have bene_id in them and b) the db has a constraint such that you cannot have a claim record without there being a beneficiary record that has the bene_id.

Reply:
I just implemented the sql queries as they were historically run, but actually looking at this...

We could get back the non-0 count results per table by just doing

```
SELECT bene_id, count(*)
FROM public.{table}
WHERE bene_id <= {start} and bene_id > {end}
GROUP BY bene_id
ORDER BY bene_id desc
```

but it doesnt give us the 0-count bene ids in the result (since those bene ids dont exist in that table), so it would require being a bit smarter about using the response (default to 0 if bene id doesnt exist in the table's response dict), unless there is some non-join sql adjustment to that query that im not thinking of that could ensure we have the full set of the requested bene ids in the result set.

Might be worth the query speed to just do some additional logic in the code when reading out of the dict results and remove a join; I'll tinker with it.

Reply:
so I think even the orig query would not get the 0-count bene_id(s). The orig did a left-join which will only return rcds where the bene_id was in both tables; using a left outer join would be necessary to find 0-count rcds.

Reply:
Was able to use the single query and a simple ternary on the place where the data is read to replace the join; good call on that! Speed was increased ~40%!

Reply:
The previous query did return the 0-count benes:
![image](https://user-images.githubusercontent.com/1200656/182199967-cd2388db-5a2a-4e29-84ea-08f857c31b72.png)
(and the new one doesnt) but due to the above I was able to use the python code to remove the need for the expensive join, so not a big deal!